### PR TITLE
Move the sourcemap filename warning.

### PR DESCRIPTION
**Pull request by chriseppstein**
The location of the sourcemap warning was tightly coupled to the calling
code. This made it impossible for an importer to have source files for
which there is no public_url without a warning from Sass.

This patch updates Sass so that the importer can disable the warning by
simply returning nil from public_url without calling
`warn_about_public_url(uri)` which is now provided by the
`Importer::Base` class.

### CODE REVIEW

File: lib/sass/engine.rb [hunk 1]
```diff
@@ -346,16 +346,24 @@ def _dependencies(seen, engines)
...
-        !(@options[:importer].public_url(@options[:filename]) ||
-          @options[:importer].is_a?(Sass::Importers::Filesystem))
+      elsif importer.nil?
         raise Sass::SyntaxError.new(<<ERR)
-Error generating source map: couldn't determine public URL for "#{@options[:filename]}".
+Error generating source map: couldn't determine public URL for "#{filename}".
+  Without a public URL, there's nothing for the source map to link to.
+  An importer was not set for this file.
+ERR
+      elsif Sass::Util.silence_warnings {!importer.public_url(filename, sourcemap_filename)}
```
Comment (nex3) [inline lib/sass/engine.rb]: Style nit: I prefer `.nil?` to `!` here.

File: lib/sass/engine.rb [hunk 2]
```diff
@@ -346,16 +346,24 @@ def _dependencies(seen, engines)
     private
 
     def _render_with_sourcemap(sourcemap_uri)
-      if @options[:filename].nil?
+      filename = @options[:filename]
+      importer = @options[:importer]
+      sourcemap_filename = @options[:sourcemap_filename] &&
```
Comment (nex3) [inline lib/sass/engine.rb]: Since this variable stores the directory containing the sourcemap, call it `sourcemap_dir`.

File: lib/sass/importers/base.rb [hunk 3]
```diff
@@ -129,8 +129,12 @@ def key(uri, options)
       # from this importer.
       #
       # @param uri [String] A URI known to be valid for this importer.
+      # @param relative_to_directory [String, NilClass] A directory (on disk) the url
+      #   should be made relative to, if possible. This directory may not be
+      #   provided.
```
Comment (nex3) [inline lib/sass/importers/base.rb]: This documentation is confusing, since in general it doesn't make sense for a server URL to be relative to a directory on disk. I suggest calling it `sourcemap_directory` and documenting it as such:

> The directory on disk where the sourcemap will be saved. If `uri` refers to a file on disk that's accessible relative to `sourcemap_directory`, this may return a relative URL.

File: lib/sass/importers/base.rb [hunk 4]
```diff
@@ -129,8 +129,12 @@ def key(uri, options)
...
       # @return [String?] The publicly-visible URL for this file.
```
Comment (nex3) [inline lib/sass/importers/base.rb]: ", or `nil` indicating that no publicly-visible URL exists."

File: lib/sass/importers/base.rb [hunk 5]
```diff
@@ -129,8 +129,12 @@ def key(uri, options)
...
-      def public_url(uri)
+      def public_url(uri, relative_to_directory = nil)
+        warn_about_public_url(uri)
```
Comment (nex3) [inline lib/sass/importers/base.rb]: I'm not a big fan of using a protected method where inheritance will do. I'd rather see the current contents of `warn_about_public_url` become the default contents of `public_url`, and have subclasses call `super` if they want to emit a warning.

File: lib/sass/importers/base.rb [hunk 6]
```diff
@@ -165,6 +169,25 @@ def directories_to_watch
...
       end
+
+      protected
+
+      # Issues a warning about being unable to determine a public url.
+      #
+      # @param uri [String] A URI known to be valid for this importer.
+      # @return [NilClass] nil
+      def warn_about_public_url(uri)
+        @warnings_issued ||= Set.new
```
Comment (nex3) [inline lib/sass/importers/base.rb]: What's the benefit of issuing a separate warning for each URI as opposed to a warning for each importer? The issue that needs to be addressed is that the importer doesn't define its own `#public_url`.

File: lib/sass/importers/filesystem.rb [hunk 7]
```diff
@@ -66,6 +66,18 @@ def watched_file?(filename)
           filename.start_with?(root + File::SEPARATOR)
       end
 
+      def public_url(name, relative_to_directory = nil)
+        if relative_to_directory.nil?
+          warn_about_public_url(name)
```
Comment (nex3) [inline lib/sass/importers/filesystem.rb:70]: I don't think the filesystem importer should warn here. I know it did prior to this patch, but it doesn't make sense to issue a warning about how a built-in importer doesn't implement a method that it actually does implement.

It should probably return an absolute `file:` URI here.

File: lib/sass/importers/filesystem.rb [hunk 8]
```diff
@@ -66,6 +66,18 @@ def watched_file?(filename)
...
+        else
+          file_path = Pathname.new(@root).join(Pathname.new(remove_root(name.to_s)))
+          file_path.relative_path_from(Pathname.new(relative_to_directory.to_s)).to_s
+        end
+      rescue ArgumentError
```
Comment (nex3) [inline lib/sass/importers/filesystem.rb]: This wraps way more code than it needs to. It should wrap _only_ the call to `#relative_path_from`. All other computation should happen outside the `begin`/`rescue` block.

Also, add a comment explaining what causes an ArgumentError.

File: lib/sass/importers/filesystem.rb [hunk 9]
```diff
@@ -66,6 +66,18 @@ def watched_file?(filename)
...
+        warn_about_public_url(name)
```
Comment (nex3) [inline lib/sass/importers/filesystem.rb]: As above, this also shouldn't warn.

File: lib/sass/importers/filesystem.rb [hunk 10]
```diff
@@ -66,6 +66,18 @@ def watched_file?(filename)
           filename.start_with?(root + File::SEPARATOR)
       end
 
+      def public_url(name, relative_to_directory = nil)
+        if relative_to_directory.nil?
+          warn_about_public_url(name)
+        else
+          file_path = Pathname.new(@root).join(Pathname.new(remove_root(name.to_s)))
+          file_path.relative_path_from(Pathname.new(relative_to_directory.to_s)).to_s
```
Comment (nex3) [inline lib/sass/importers/filesystem.rb]: `Pathname#relative_path_from` will break if one path is absolute and the other is relative. Are we guaranteeing that `relative_to_directory` is always absolute? If so, document it. If not, make it absolute here.

Also, calling `to_s` on `relative_to_directory` here and `name` above is unnecessary.

File: lib/sass/importers/filesystem.rb [hunk 11]
```diff
@@ -66,6 +66,18 @@ def watched_file?(filename)
           filename.start_with?(root + File::SEPARATOR)
       end
 
+      def public_url(name, relative_to_directory = nil)
+        if relative_to_directory.nil?
+          warn_about_public_url(name)
+        else
+          file_path = Pathname.new(@root).join(Pathname.new(remove_root(name.to_s)))
```
Comment (nex3) [inline lib/sass/importers/filesystem.rb]: Just call `File.absolute_path(name)`.

File: test/sass/importer_test.rb [hunk 12]
```diff
@@ -213,10 +225,58 @@ def test_source_map_with_only_css_uri_supports_public_url_imports
 JSON
   end
 
+  def test_source_map_with_only_css_uri_can_have_no_public_url_without_warning
+    $TEST = true
```
Comment (nex3) [inline test/sass/importer_test.rb]: Left over from debugging?

File: test/sass/scss/scss_test.rb [hunk 13]
```diff
@@ -1650,42 +1650,6 @@ def test_function_triple_keyword_splat_merge
...
-  @return $str;
-}
-
-.foo {
-  $map: (a-1: 1, b_2: 2, c-3: 3, d_4: 4);
-  val: foo($map...);
-}
-SCSS
-  end
-
```
Comment (nex3) [inline test/sass/scss/scss_test.rb:1838]: Why is this deleted?

File: test/test_helper.rb [hunk 14]
```diff
@@ -37,6 +38,15 @@ def filename_for_test(syntax = :sass)
     "#{test_name}_inline.#{syntax}"
   end
 
+  def sourcemap_filename_for_test(syntax = :sass)
+    test_name = caller.
+      map {|c| Sass::Util.caller_info(c)[2]}.
+      compact.
+      map {|c| c.sub(/^(block|rescue) in /, '')}.
+      find {|c| c =~ /^test_/}
```
Comment (nex3) [inline test/test_helper.rb]: Abstract out this `test_name` computation.

File: test/test_helper.rb [hunk 15]
```diff
@@ -37,6 +38,15 @@ def filename_for_test(syntax = :sass)
...
+    "#{test_name}.css.map"
```
Comment (nex3) [inline test/test_helper.rb]: Add "_inline" before ".css".

**Comment by nex3** [comment]
Done reviewing.

File: test/sass/scss/scss_test.rb (hunk 13 above)
Comment (chriseppstein) [inline test/sass/scss/scss_test.rb:1838]: These same tests were in the code twice.

File: lib/sass/importers/filesystem.rb (hunk 10 above)
Comment (nex3) [inline lib/sass/importers/filesystem.rb]: One more thing: we can fix #839 by calling `Pathname#cleanpath` on both paths here before calling `#relative_path_from`.

File: lib/sass/engine.rb (hunk 2 above)
Comment (chriseppstein) [inline lib/sass/engine.rb]: Changed.

File: lib/sass/engine.rb (hunk 1 above)
Comment (chriseppstein) [inline lib/sass/engine.rb]: Changed.

File: lib/sass/importers/base.rb (hunk 6 above)
Comment (chriseppstein) [inline lib/sass/importers/base.rb]: There are two issues at play. The first is that the importer may not have defined what a public_url means. The second issue is that even though the semantics of a public_url for filesystems is fairly well defined, it's possible for there to be no way to construct the public url for the filesystem paths in use. E.g. If the sourcemap is being generated into a different disk (on windows) than the stylesheet. In this case, the per-uri warning makes sense to me. I've separated this into two warnings. There's the default warning about implementing the method, but there's also a warning in the filesystem importer.

File: lib/sass/importers/base.rb (hunk 5 above)
Comment (chriseppstein) [inline lib/sass/importers/base.rb]: `warn_about_public_url` is moved to the filesystem importer now.

File: lib/sass/importers/filesystem.rb (hunk 7 above)
Comment (chriseppstein) [inline lib/sass/importers/filesystem.rb:70]: >  It should probably return an absolute `file:` URI here.

This should be done in a different pull request. It's unrelated to this change.

File: lib/sass/importers/base.rb (hunk 4 above)
Comment (chriseppstein) [inline lib/sass/importers/base.rb]: Changed.

File: lib/sass/importers/base.rb (hunk 3 above)
Comment (chriseppstein) [inline lib/sass/importers/base.rb]: Changed.

File: lib/sass/importers/filesystem.rb (hunk 11 above)
Comment (chriseppstein) [inline lib/sass/importers/filesystem.rb]: I don't understand how that's useful in this context.

File: lib/sass/importers/filesystem.rb (hunk 10 above)
Comment (chriseppstein) [inline lib/sass/importers/filesystem.rb]: Yes. It's always absolute. I've documented it and refactored this method and added cleanpath calls.

File: lib/sass/importers/filesystem.rb (hunk 8 above)
Comment (chriseppstein) [inline lib/sass/importers/filesystem.rb]: Moved. Comment added.

File: lib/sass/importers/filesystem.rb (hunk 9 above)
Comment (chriseppstein) [inline lib/sass/importers/filesystem.rb]: It should. As noted above.

File: test/sass/importer_test.rb (hunk 12 above)
Comment (chriseppstein) [inline test/sass/importer_test.rb]: Yep. Removed.

File: test/test_helper.rb (hunk 15 above)
Comment (chriseppstein) [inline test/test_helper.rb]: added.

File: test/test_helper.rb (hunk 14 above)
Comment (chriseppstein) [inline test/test_helper.rb]: refactored.

File: lib/sass/importers/filesystem.rb (hunk 11 above)
Comment (nex3) [inline lib/sass/importers/filesystem.rb]: `Pathname.new(@root).join(Pathname.new(remove_root(name.to_s)))` is effectively making `name` absolute, but it's doing so in a very roundabout way that runs the risk of failing if `name`'s prefix isn't textually identical to `@root`'s.

File: lib/sass/importers/base.rb [hunk 16]
```diff
@@ -125,8 +125,18 @@ def key(uri, options)
       # from this importer.
       #
       # @param uri [String] A URI known to be valid for this importer.
-      # @return [String?] The publicly-visible URL for this file.
-      def public_url(uri)
+      # @param sourcemap_directory [String, NilClass] The absolute path to a
+      #   directory on disk where the sourcemap will be saved. If uri refers to
+      #   a file on disk that's accessible relative to sourcemap_directory, this
+      #   may return a relative URL.
```
Comment (nex3) [inline lib/sass/importers/base.rb]: Add "This may be `nil` if the sourcemap's eventual location is unknown."

File: lib/sass/importers/base.rb [hunk 17]
```diff
@@ -125,8 +125,18 @@ def key(uri, options)
...
+      # @return [String?] The publicly-visible URL for this file, or `nil`
+      #   indicating that no publicly-visible URL exists.
+      def public_url(uri, sourcemap_directory = nil)
+        return if @public_url_warning_issued
+        @public_url_warning_issued = true
+        Sass::Util.sass_warn <<WARNING
```
Comment (nex3) [inline lib/sass/importers/base.rb:138]: It looks like there are no tests that check for this warning. We should add one.

File: lib/sass/importers/base.rb (hunk 6 above)
Comment (nex3) [inline lib/sass/importers/base.rb]: Sounds good to me.

**Comment by nex3** [comment]
Done reviewing. There are also a couple replies in the inline comments from the last iteration.

File: lib/sass/importers/base.rb (hunk 17 above)
Comment (chriseppstein) [inline lib/sass/importers/base.rb:138]: That would cause an error instead of a warning.

File: lib/sass/importers/base.rb (hunk 16 above)
Comment (chriseppstein) [inline lib/sass/importers/base.rb]: Added.

**Comment by chriseppstein** [comment]
Squashed and merged to master as f9a615c1e3ab0758976b438b3244487b3249c4d7.
